controllers/hostingdeploymentautoscalingjob/hostingdeploymentautoscalingjob_controller.go

```diff
@@ -0,0 +1,449 @@
+		if err = r.deleteAutoscalingPolicy(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to deleteAutoscalingPolicy"))
+		}
+
+		// Delete succeeded, set Description to nil.
+		ctx.ScalingPolicyDescriptionList = nil
+		ctx.ScalableTargetDescriptionList = nil
+	}
+
+	// If update or create, create the desired hostingdeploymentautoscaling.
+	if action == controllers.NeedsCreate || action == controllers.NeedsUpdate {
+		if err = r.applyAutoscalingPolicy(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to applyAutoscalingPolicy"))
+		}
+	}
+
+	// Update the status accordingly.
+	// Review: Deleted Status may be unnecessary
+	status := CreatedAutoscalingJobStatus
+	if ctx.ScalingPolicyDescriptionList == nil {
+		status = DeletedAutoscalingJobStatus
+	}
+	if err = r.updateStatus(ctx, status); err != nil {
+		return err
+	}
+
+	// Remove the Finalizer on delete
+	if controllers.HasDeletionTimestamp(ctx.HostingDeploymentAutoscalingJob.ObjectMeta) {
+		if err = r.removeFinalizer(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Removes the finalizer held by our controller.
+func (r *Reconciler) removeFinalizer(ctx reconcileRequestContext) error {
+	var err error
+
+	ctx.HostingDeploymentAutoscalingJob.ObjectMeta.Finalizers = controllers.RemoveString(ctx.HostingDeploymentAutoscalingJob.ObjectMeta.Finalizers, controllers.SageMakerResourceFinalizerName)
+	if err = r.Update(ctx, ctx.HostingDeploymentAutoscalingJob); err != nil {
+		return errors.Wrap(err, "Failed to remove finalizer")
+	}
+	ctx.Log.Info("Finalizer has been removed")
+
+	return err
+}
+
+// getResourceIDListfromInputSpec converts the list of resources into a string list to be used for various API calls
+func (r *Reconciler) getResourceIDListfromInputSpec(ctx *reconcileRequestContext) error {
+
+	resourceIDListfromSpec := ctx.HostingDeploymentAutoscalingJob.Spec.ResourceID
+	for _, resourceIDfromSpec := range resourceIDListfromSpec {
+		ResourceID := sdkutil.ConvertAutoscalingResourceToString(*resourceIDfromSpec)
+		ctx.ResourceIDList = append(ctx.ResourceIDList, *ResourceID)
+	}
+
+	// TODO: error handling
+	return nil
+}
+
+// Initialize fields on the context object which will be used later.
+func (r *Reconciler) initializeContext(ctx *reconcileRequestContext) error {
+	var err error
+
+	// Ensure we are using the job name specified in the spec
+	if ctx.HostingDeploymentAutoscalingJob.Spec.PolicyName != nil && len(*ctx.HostingDeploymentAutoscalingJob.Spec.PolicyName) > 0 {
+		ctx.PolicyName = *ctx.HostingDeploymentAutoscalingJob.Spec.PolicyName
+	} else {
+		ctx.PolicyName = DefaultAutoscalingPolicyName
+		ctx.HostingDeploymentAutoscalingJob.Spec.PolicyName = &ctx.PolicyName
+
+		// Review: Is this needed
+		if err := r.Update(ctx, ctx.HostingDeploymentAutoscalingJob); err != nil {
```

Comment:
There is no generated policy name in this case since that's not supposed to be unique. What we should add is the policy ARN. But the problem with that is that there will be as many as the number of variants. to be discussed with Alex whether this should be added to HAP or HD